src/vdom/index.js

```diff
@@ -10,19 +12,38 @@ export function isSameNodeType(node, vnode, hydrating) {
 	if (typeof vnode==='string' || typeof vnode==='number') {
 		return node.splitText!==undefined;
 	}
-	if (typeof vnode.nodeName==='string') {
-		return !node._componentConstructor && isNamedNode(node, vnode.nodeName);
+	const { _componentConstructor } = node;
+	const { nodeName } = vnode;
+	if (typeof nodeName==='string') {
+		return !_componentConstructor && isSameNodeName(node, nodeName);
 	}
-	return hydrating || node._componentConstructor===vnode.nodeName;
+	return hydrating || _componentConstructor===nodeName || isSameNodeConstructor(node, nodeName);
 }
 
 
 /** Check if an Element has a given normalized name.
-*	@param {Element} node
-*	@param {String} nodeName
+ *	@param {Element} node
+ *	@param {mixed} nodeName
+ */
+export function isSameNodeName(node, nodeName) {
+	return node.normalizedNodeName===nodeName || lowercase(node.nodeName)===lowercase(nodeName);
+}
+
+
+/** Check if a node is the same via constructor.
+ *  @param {Element} node
+ *  @param {Function} nodeName
+ */
+export function isSameNodeConstructor({ constructor }, nodeName) {
```

Comment:
i know this sounds crazy, but preact doesn't contain any destructuring as a rule. it makes the output larger and hides a property access that often need be safeguarded. In this particular case it's likely a lot faster to inline `.constructor===nodeName` rather than pass many object types to a test function.